lib/Reference/WorkPackageReferenceProvider.php

```diff
@@ -0,0 +1,188 @@
+ */
+
+namespace OCA\OpenProject\Reference;
+
+use OCA\OpenProject\Service\OpenProjectAPIService;
+use OCP\Collaboration\Reference\ADiscoverableReferenceProvider;
+use OCP\Collaboration\Reference\ISearchableReferenceProvider;
+use OCP\Collaboration\Reference\Reference;
+use OC\Collaboration\Reference\ReferenceManager;
+use OCA\OpenProject\AppInfo\Application;
+use OCP\Collaboration\Reference\IReference;
+use OCP\IConfig;
+use OCP\IL10N;
+use OCP\IURLGenerator;
+
+class WorkPackageReferenceProvider extends ADiscoverableReferenceProvider implements ISearchableReferenceProvider {
+	private const RICH_OBJECT_TYPE = Application::APP_ID . '_work_package';
+
+	// as we know we are on NC >= 26, we can use Php 8 syntax for class attributes
+	public function __construct(private IConfig $config,
+								private IL10N $l10n,
+								private IURLGenerator $urlGenerator,
+								private ReferenceManager $referenceManager,
+								private OpenProjectAPIService $openProjectAPIService,
+								private ?string $userId) {
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getId(): string {
+		return 'openproject-work-package-ref';
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getTitle(): string {
+		return $this->l10n->t('OpenProject work packages');
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getOrder(): int {
+		return 10;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getIconUrl(): string {
+		return $this->urlGenerator->getAbsoluteURL(
+			$this->urlGenerator->imagePath(Application::APP_ID, 'app-dark.svg')
+		);
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function getSupportedSearchProviderIds(): array {
+		if ($this->userId !== null) {
+			$ids = [];
+			// take the fallback value from the defaults
+			$searchEnabled = $this->config->getUserValue(
+				$this->userId,
+				Application::APP_ID,
+				'search_enabled',
+				$this->config->getAppValue(Application::APP_ID, 'default_enable_unified_search', '0')
+			);
+			if ($searchEnabled) {
+				$ids[] = 'openproject-search';
+			}
+			return $ids;
+		}
+		return ['openproject-search'];
+	}
+
+	/**
+	 * Parse a link to find a work package ID
+	 *
+	 * @param string $referenceText
+	 *
+	 * @return int|null
+	 */
+	private function getWorkPackageIdFromUrl(string $referenceText): ?int {
+		// example links
+		// https://community.openproject.org/projects/nextcloud-integration/work_packages/40070
+		$openProjectUrl = $this->config->getAppValue(Application::APP_ID, 'openproject_instance_url');
+		preg_match('/^' . preg_quote($openProjectUrl, '/') . '\/projects\/[^\/\?]+\/work_packages\/([0-9]+)/', $referenceText, $matches);
+		if (count($matches) > 1) {
+			return (int) $matches[1];
+		}
+
+		return null;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function matchReference(string $referenceText): bool {
+		if ($this->userId !== null) {
+			$linkPreviewEnabled = $this->config->getUserValue($this->userId, Application::APP_ID, 'default_enable_unified_search', '0') === '1';
```

Comment:
reminder to self to change these and  add a migration step to add this key
```suggestion
			$linkPreviewEnabled = $this->config->getUserValue($this->userId, Application::APP_ID, ''link_preview_enabled'', '0') === '1';
```